Delta / Change based meeting Imports

gkovats opened the pull request:
This is basically a rewrite of #1398. 

- There's a common import sanitization method that reuses a lot of the current logic when meetings are imported
- It adds a few new fields, which are used for checking for later changes from the import source
  - source_slug
  - source_region
  - source_sub_region
  - source_formatted_address
- Imports and change detect emails now use a common method to only return details / import changed or new meetings
- Imports now perform updates to existing posts, so instead of creating brand new posts each import, it's just attempting to connect the imported meeting with the previous and perform updates

gkovats commented on includes/ajax.php:
```diff
@@ -393,15 +393,24 @@ function tsml_ajax_geocodes()
     //get lookups, todo consider adding regions to this
     $locations = $groups = [];
     $all_locations = tsml_get_locations();
```
I just cleaned up some of the formatting here to better read it. No code change.

gkovats commented on includes/ajax.php:
```diff
@@ -393,15 +393,24 @@ function tsml_ajax_geocodes()
             $errors[] = '<li value="' . $meeting['row'] . '">' . sprintf(__('No location information provided for <code>%s</code>.', '12-step-meeting-list'), $meeting['name']) . '</li>';
             continue;
         }
 
```
Rather than call them out in 4 spots, I put the new fields `source_*` in $tsml_source_fields_map for use in different spots. They won't be meaningful for in local, non-imported meetings, just internal fields for mapping meetings we pulled in to ones coming from the import source.

gkovats commented on includes/functions.php:
```diff
@@ -108,6 +108,42 @@ function tsml_bounds()
     }
 }
 
```
This method expects an array based log of changes from `tsml_get_changed_import_meetings()` to spit out a table. That's it. Just so we don't have to cobble a table together in the meeting change detection code.

kiyote33 replied:
Here's the table report generated when tsml_debug is true:
![image](https://github.com/user-attachments/assets/7e31c43f-1b02-443f-84e4-51c1a30c6249)

Here's the table report we had before: 
![image](https://github.com/user-attachments/assets/b6e503cb-5648-4a27-b8d5-733ee39b7540)
The new report has the Header text centered in each column. Think the new report might be better if left aligned and delineated by grid lines so it's easier to read.

gkovats replied:
I'm kinda up the in air about the Notes column for starters, not sure if folks would _want_ to see the system names for the fields changed. I agree the header columns should be left aligned, will add that. I think the colors kinda throw it off, and the table border looks a little odd, even if it adds separation. I'll tweak the header cols to left align and try that.

gkovats commented on includes/functions.php:
```diff
@@ -1205,13 +1239,21 @@ function tsml_get_meetings($arguments = [], $from_cache = true, $full_export = f
             if (!empty($meeting_meta[$post->ID]['data_source']) && !empty($tsml_data_sources[$meeting_meta[$post->ID]['data_source']]['name'])) {
                 $meeting['data_source_name'] = $tsml_data_sources[$meeting_meta[$post->ID]['data_source']]['name'];
             }
 
```
Just adding those new `source_*` fields

gkovats commented on includes/functions.php:
```diff
@@ -1471,11 +1518,14 @@ function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source
     $user_id = get_current_user_id();
 
```
Since Google spreadsheets can have entirely non-unique slugs or none at all, so we can't really tie incoming meetings to ones in the DB. So just made certain in this step I had a semi-reliable, unique slug, so I could better connect incoming to ones we have.

gkovats commented on includes/functions.php:
```diff
@@ -1471,11 +1518,14 @@ function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source
-            $item = utf8_encode($item);
-        }
-    });
+    if (function_exists('mb_detect_encoding')) {
```
I think I didn't have this PHP lib installed an so, just wrapping with a check. Or maybe Wayne had this in there. I don't know anymore, Git blame should know.

joshreisner replied:
we did recently encounter a user who didn't have the `mb_` functions enabled and started checking

kiyote33 replied:
Hmmm. I don't remember adding it. Think it came in on a merge after a patch release, which aligns with what Josh said.

gkovats commented on includes/functions.php:
```diff
@@ -1497,6 +1547,33 @@ function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source
     $indexes_to_remove = [];
 
     for ($i = 0; $i < count($meetings); $i++) {
+
```
Since Google spreadsheets can have entirely non-unique slugs or none at all, so we can't really tie incoming meetings to ones in the DB. So just made certain in this step I had a semi-reliable, unique slug, so I could better connect incoming to ones we have.

kiyote33 replied:
That was the reason we decided not to import the google spreadsheets as updateable data sources. Are we now doing so?

gkovats replied:
Yep, we figured really to treat all data sources the same way for simplification. The sanitize method does still call our existing speadsheet cleanup function, but otherwise this does an adequate job of coming up with consistent slugs to use, and with that we can compare old to new.

kiyote33 replied:
Sounds good...

gkovats commented on includes/functions.php:
```diff
@@ -2426,7 +2421,154 @@ function tsml_date_localised($format, $timestamp = null)
     return $datetime->format($format);
 }
-/* ******************** end of data_source_change_detection ******************** */
+
```
In effect this new method 
- tries to map existing meetings to the incoming ones
- for matches, check for changes
- return 3 arrays, documented below

kiyote33 commented:
Hi George, Two items observed in the following image. On refresh the initial message about when the next cron job is scheduled to run doesn't need to display every time the refresh button is clicked. More important though is the number of meetings is zeroed out. That needs to be kept or refreshed. I'll start in-depth testing tomorrow. Your code re-write looks great!

joshreisner commented on includes/functions.php:
```diff
@@ -1497,6 +1547,33 @@ function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source
+            $meeting_slugs[$meeting_slug] = 1;
+        }
+        $meetings[$i]['slug'] = $meeting_slug;
+
```
cool! so the logic here is that it tries to use the above series of fields, and sanitizes then and adds a number 1-999 to slugify the record? 

not sure if we're doing this already but GSO instructed some google sheets to use a field called `ID` in lieu of slug, so we should support that

also `post_name` is a field that wordpress uses in their db but i don't think we expect that in JSON, CSV, or in a sheet

gkovats replied:
Fair, likely if `post_name` exists, `slug` would as well. I was largely doing that for overkill - just exhaustively trying to find all valid fields. I'll add `ID`, but will reject it if it's a number like a WordPress post ID.

joshreisner commented on includes/admin_import.php:
```diff
@@ -164,31 +165,61 @@ function tsml_import_page()
                     'status' => 'OK',
                     'last_import' => current_time('timestamp'),
-                    'count_meetings' => 0,
+                    'count_meetings' => count($import_meetings),
```
nbd but i wonder if this is too hasty now to be setting  the finished number of meetings? should we do this instead after the real number has been modified, in case something goes wrong?

gkovats replied:
Yeah, I changed this yesterday when I saw the number wasn't updating in the Ajax step. I can move it there, it'll just have to query all source meetings for an accurate count after each AJAX import run.

joshreisner reviewed (approved):
code looks good, thanks for this great effort @gkovats and thank you @kiyote33 as well for contributing the foundation! provisionally approving - haven't had a chance to test it yet and will be away for a week, but if @kiyote33 also approves, happy to have this go to main

gkovats commented:
I'll handle the code conflict, and @kiyote33 I'll follow up on whatever you find. That ajax cron message, I'll look into what caused that and address.

kiyote33 reviewed (commented):
This test checks a hybrid meeting which is being reset to in-person only. The import record has no values for conference url, conference notes, conference phone, or conference phone notes. The matching database record has values for all those, so null values for all those attributes should be entered in the update but that doesn't happen. Here's the report: 
![image](https://github.com/user-attachments/assets/8855e564-bd00-405c-bd18-0610c9f405c8)
Test feed URL: https://[eco.parkercanada.info/wp-admin/admin-ajax.php?action=meetings](https://eco.parkercanada.info/wp-admin/admin-ajax.php?action=meetings)
PS... I think the report would be better served if we replaced the Group Name attribute/value with the Slug attribute/value.

gkovats commented:
Updates:
- added `ID` to source fields for the meeting slug, _but_ only if it's non-numeric. I'm kinda paranoid, `ID` could easily be some integer they use to track meetings on their end... though, guess that's a valid slug too? Meh, tell me if that's overkill
- updated the AJAX import to update Group and Location notes
- updated how the sources are updated on initial scrape and loading into options - existing sources, I just update the `last_import` timestamp, leave the count in place. The AJAX loop updates that total count as it chews through batches.
- added Slug to that debug table for import changes, as well as left aligned the table header and added some `cellpadding`. I had to look that attribute up on MDN because it's been 10 years since I used it last. The Web is wild, man.

gkovats commented on includes/functions.php:
```diff
@@ -2601,6 +2601,13 @@ function tsml_compare_imported_meeting($local_meeting, $import_meeting)
         }
         $normalized_meetings[] = $normalized_meeting;
     }
+
```
don't compare group fields when no group present

kiyote33 replied:
Testing Failed. Using https://[sheets.googleapis.com/v4/spreadsheets/13W4lBuRWKpnHNOC_3wTXI5anLVOkyvMmyn4Wvqx1z3c/values/A1:ZZ?key=AIzaSyAdMoeqxk-EWspUIJDwN7GBSwWSzgL6E5A](https://sheets.googleapis.com/v4/spreadsheets/13W4lBuRWKpnHNOC_3wTXI5anLVOkyvMmyn4Wvqx1z3c/values/A1:ZZ?key=AIzaSyAdMoeqxk-EWspUIJDwN7GBSwWSzgL6E5A) for Puget Sound.

gkovats commented on includes/functions.php:
```diff
@@ -131,7 +131,7 @@ function tsml_build_import_change_report($data_source_name, $change_log)
         $meeting_day_val = intval(isset($meeting['day']) ? $meeting['day'] : -1);
         $meeting_day = isset($tsml_days[$meeting_day_val]) ? $tsml_days[$meeting_day_val] : '?';
         $meeting_time = isset($meeting['time']) ? $meeting['time'] : '?';
```
adjusted value to "No group" since that's a valid state for meetings

gkovats commented on includes/admin_import.php:
```diff
@@ -195,6 +195,11 @@ function tsml_import_page()
                         'change_detect' => $data_source_change_detect,
                         'type' => 'JSON',
                     ];
```
Moved line to clause where new feed is added, rather than to be run every import sync

kiyote33 replied:
Excellent. Only shows when a new data source is added.

gkovats commented on includes/ajax.php:
```diff
@@ -529,7 +530,9 @@ function tsml_ajax_geocodes()
             // timezone
             if (!empty($meeting['timezone']) && in_array($meeting['timezone'], DateTimeZone::listIdentifiers())) {
                 update_post_meta($location_id, 'timezone', $meeting['timezone']);
```
Delete timezone when blank

gkovats commented on includes/functions.php:
```diff
@@ -2578,7 +2577,7 @@ function tsml_compare_imported_meeting($local_meeting, $import_meeting)
         array_keys($tsml_export_columns),
         //these fields are unique internal fields, not content fields for comparison
-        explode(',', 'id,name,slug,author,data_source,data_source_name')
+        explode(',', 'id,slug,author,data_source,data_source_name')
```
Compare name field, and fallback to `post_title` when name not present

kiyote33 reviewed (approved):
My testing is complete with all issues resolved. Thanks for your hard work George!!!
